File: 0299_Problem_1.jl
# Problem Link: https://leetcode.com/contest/weekly-contest-299/problems/check-if-matrix-is-x-matrix/

using LinearAlgebra # I

# Using I (for Identity)
function checkmatrix(grid)
  i = size(grid, 1) |> I |> Matrix # identity matrix
  min.(grid, 1) == max.(i, reverse(i, dims=1))
end

# Without using I
function checkmatrix(grid)
  n = size(grid, 1)
  i = 1:n .== 1:n |> transpose
  min.(grid, 1) == max.(i, reverse(i, dims=1))
end

# Tests
checkmatrix([2 0 0 1; 0 3 1 0; 0 5 2 0; 4 0 0 2]) # true
checkmatrix([5 7 0; 0 3 1; 0 5 0])                # false

File: 0299_Problem_1.py
# Problem Link: https://leetcode.com/contest/weekly-contest-299/problems/check-if-matrix-is-x-matrix/

import numpy as np

# Using np.identity
def check_matrix(grid):
    i = np.identity(len(grid), int)
    return np.array_equal(np.minimum(grid, 1), np.maximum(i, np.flip(i, axis=1)))

# Without using np.identity
def check_matrix(grid):
    n = len(grid)
    i = np.equal.outer(np.arange(n), np.arange(n))
    return np.array_equal(np.minimum(grid, 1), np.maximum(i, np.flip(i, axis=1)))

check_matrix(np.matrix('2 0 0 1; 0 3 1 0; 0 5 2 0; 4 0 0 2')) # True
check_matrix(np.matrix('5 7 0; 0 3 1; 0 5 0'))                # False
